### DISCUSSION TITLE
Get Normal vector / the face of a mesh containing a point

### DISCUSSION THREAD

**Discussion by Yarkane** [jimy-byerley/pymadcad#45] [Q&A] [answered]
Hi there, that's me again, hoping I don't get too much of your time !

This time I have a simple question and I don't see the answer anywhere in the documentation.

I need to get the normal vector of a mesh surface in a given point, for example the point I clicked.

I found, in the documentation, the function `facenormal(f)` : https://pymadcad.readthedocs.io/en/latest/reference/mesh.html?highlight=normal#madcad.mesh.Mesh.facenormal
This function gives me the output I need : the normal vector of a face.

But I lack one last brick to make it happen : to get the face I'm clicking on when I click on a mesh.
I don't find any function allowing me to do it, I know i can call mesh.faces() to get all the faces, but that does not fullfil my need.

So I would have something like this : `mesh.facenormal(mesh.???(self.ptat(evt.pos())))`

Do you have a tip on this one ?

Thanks again for your precious time ! 😄

**Comment by jimy-byerley** [comment, marked as answer]
Hello @Yarkane !
Unfortunately, there is no such thing out of the box in madcad. 

The problem of getting the index of a face intersecting a ray (from the view point to the clicked direction for instance) is a common problem, typically in rendering programs featuring ray tracing, or in collision detection. But those are not the main purpose of madcad (there is nothing wrong in using madcad together with such functions, but that's not a target for madcad).

In the madcad rendering pipeline, the problem is addressed only for the purpose of mouse-selection in the GUI, and implemented in a way that only works for GUI interactions:  every frame rendered in the window also renders an *identification* frame where each pixel value is the ident of the object in the scene that is below this pixel. (it basically is [rasterization](https://en.wikipedia.org/wiki/Rasterisation), made on the GPU)

- For now, the ident are refering to [mesh groups](https://pymadcad.readthedocs.io/en/latest/reference/mesh.html#madcad.mesh.Mesh.groups), allowing the user to select a determined portion of surface by one click. to get the group index at some position on the screen, you can use [`View.itemat()`](https://pymadcad.readthedocs.io/en/latest/reference/rendering.html#madcad.rendering.View.itemat) which returns the name of the mesh selected followed by its seleced group index.
- With small modifications, the madcad rendering pipeline could rasterize idents for the faces (face indices) instead of (or in addition to) idents for groups (groups indices). 

  The modification is as simple as this
  ```python
  return displays.SolidDisplay(scene, 
		typedlist_to_numpy(m.points, 'f4'), 
		typedlist_to_numpy(normals, 'f4'), 
		typedlist_to_numpy(m.faces, 'u4'),
		typedlist_to_numpy(edges, 'u4'),
		idents = np.arange(len(faces), dtype='u4'),  # changed line
		color = self.options.get('color'),
		)
   ```
  Plus some modifications in the event handling of `SolidDisplay` so that idents retreived are no more treated as groups 

  Group selection is more likely to serve the end-user of madcad than facet selection (and will likely become increasingly important  with the coming features), so I will keep the rendering pipeline the way it is. But you can for your own decide to change the way idents are used by overriding the [display for `Mesh`](https://github.com/jimy-byerley/pymadcad/blob/50e9971e2fc763dcab0d4aa221c863ea6753727a/madcad/mesh.py#L907), so it will rasterize facet indices instead of group indices. 

- Alternatively, you can just make mesh groups match facets (nothing special is necessary to do it, but it will render with wireframe everywhere on every triangles)
  ```python
  mymesh = Mesh(...)  # the original mesh
  show([mymesh])   # renders nicely, and clicks are providing group indices
  
  second = Mesh(
            mymesh.points, 
            mymesh.faces, 
            tracks = range(len(mymesh.faces)), 
            groups = mymesh.faces,
            )  # faces tracks refers to groups which are faces ... nice brainfuck !
  show([second])   # renders with edges between every triangles, but clicks are providing facets indices 
  ```

I hope this help !

**Reply by Yarkane** [reply]
Thanks a lot ! I'll take time to think about these solutions.
